### PR TITLE
Object selection

### PR DESCRIPTION
**Pull request by lethosor**
## What changes are proposed in this pull request?

Initial support for selecting individual objects in the media viewer

image

Selecting all objects in a frame is also supported for videos. More selection options are in progress.

## How is this patch tested? If it is not, please explain why.

Locally

### Is this a user-facing change that should be mentioned in the release notes?

-   [ ] No. You can skip the rest of this section.
-   [x] Yes. Give a description of this change to be included in the release
        notes for FiftyOne users.

### What areas of FiftyOne does this PR affect?

-   [x] App: FiftyOne application changes
-   [ ] Build: Build and test infrastructure changes
-   [ ] Core: Core `fiftyone` Python library changes
-   [ ] Documentation: FiftyOne documentation changes
-   [ ] Other

### CODE REVIEW

**Comment by lethosor** [comment]
@brimoor is there a reason that you keep pushing to this branch? I'm not seeing any new changes from you that aren't already on develop, and since I'm still working on this branch, I'd prefer to avoid having to merge your merges.

**Comment by brimoor** [comment]
I merged develop once because https://github.com/voxel51/fiftyone/pull/638 was merged in there. "keep merging" is a bit sensational.

**Comment by lethosor** [comment]
I already took care of the merge in 9863593137e911c892b31b6fada65bca7e022413. My comment was in response to 51bf6bf579268de37234c490be995a25d0eb4a9a, which is the second time you've pushed to this branch (I undid the first). 

I'm fine with keeping the merge this time, and I understand your reasoning behind it, but it creates extra work for me when I attempt to push my changes, which is why I'd prefer that I handle these sorts of updates on branches that I create.

**Comment by benjaminpkane** [comment]
I'm seeing a server error when hovering over a video in the quickstart:
```
Server Error

An exception has been raised by the server. Your session has been reverted to its previous state.

Traceback (most recent call last):
  File "/home/ben/code/fiftyone/fiftyone/server/main.py", line 109, in wrapper
    return func(self, *args, **kwargs)
  File "/home/ben/code/fiftyone/fiftyone/server/main.py", line 335, in on_get_frame_labels
    field_labels = _make_frame_labels(k, v, frame_number)
  File "/home/ben/code/fiftyone/fiftyone/server/main.py", line 169, in _make_frame_labels
    obj._id = str(label["_id"])
KeyError: '_id'
```

Looks like it is trying to grab the id of `Detections` and not a `Detection`.

**Comment by lethosor** [comment]
Thanks, I'll look into it - I thought I was checking for list fields appropriately, but that logic may have needed to change after a merge.

**Comment by lethosor** [comment]
Looks like that path is only hit for `Detections` and other list fields on videos (since images are handled on the frontend). I do have access to both the FiftyOne and ETA labels in _make_frame_labels(), but matching them up at this point is challenging.

Before I attempt to modify `ImageLabel`-derived classes to make this easier, I have a couple questions, just to make sure I'm not missing something that exists already:
* Is there currently a way to identify label types that contain a list of other labels? i.e. a way to distinguish (Classifications, Detections, Keypoints, Polylines) from (Classification, Detection, Keypoint, Polyline)
* Is there currently a way to access all labels contained in a list type (from above) independent of its name? i.e. accessing `Classifications().classifications`, `Detections().detections`, etc. without needing a special case for each field name. I am reluctant to rely on `type(label).__name__.lower()` because that might not always work.

**Comment by benjaminpkane** [comment]
I cover both of those points in #631 . Unfortunately, I am assuming `type(label).__name__.lower()` for the attribute. Using a mixin is probably best, long term.

I would recommend just using the code snippet below, I can react to you in #631.

```py
_LABELS = (fol.Classifications, fol.Detections, fol.Keypoints, fol.Polylines)

def _get_items(label):
    return getattr(label, type(label).__name__.lower())
```

**Comment by lethosor** [comment]
Ok, thanks. My current approach in _make_image_labels is a bit temporary anyway, so I may as well use that or search for attributes that are lists. I was also thinking of a mixin as a longer-term solution (maybe `_HasList` with a `_list_field` property pointing to the relevant field?) but I won't go down that road if it'll conflict with #631.

**Comment by brimoor** [comment]
Works great! I like the way you handled the selection/deselection options at both the sample and dataset levels.

From a usability standpoint, here my suggestions for this awesome feature:

- Since object selection applies only to labels, I think the right location for the `X object selected` element is probably closer to the `LABELS` tab on the sidebar. The design principle being to keep semantically related things close together. If it looks funny when placed above the first label field or below the last one, then perhaps the next best location is underneath `Display Options`, so that it is as close as possible

- Make the selection area for objects be the boundary of the box (including the label header), not the interior of the box. The reasoning is that, for most realistic scenes, there will be many overlapping boxes and making the box interiors selectable basically guarantees that many boxes will never be selectable. The same comment would apply to keypoints and polylines as well. Basically, the target is "if something was rendered on the canvas, you can click it; otherwise, you can't" (I noticed when testing with BDD100k that polyline selection works too; awesome!)

- Make "Select all" and "Unselect all" only apply to the labels that are currently shown in the media player. For example, if there are 3 label fields and two are deselected and the other has a confidence filter applied, only select/deselect the currently visible object, not the hidden fields/objects. This is similar in spirit to the first point; the mindset is that actions apply to what you see, and not things that you don't see

**Comment by lethosor** [comment]
Yeah, I guess this does only apply to things in the "Labels" selection currently. Adding things in the middle of the sidebar is a bit tricky. Under "display options" might be a good compromise for now.

> Make the selection area for objects be the boundary of the box..., not the interior of the box

This is just a 1-2px difference, right? The current logic uses the actual coordinates of the object, while the line drawn around it may extend outside of the object a bit. This behavior might be possible to change, but I could add a tolerance of `line width / 2` to the boundary calculation fairly easily (I think), if that would do what you want.

>  (including the label header)

Clicking on the title of a detection (e.g. "DOG" above) works for me - does it not work for you? Again, might be subject to a 1-2px difference depending on whether that header also has a border drawn.

**Comment by brimoor** [comment]
I'm pretty confident that users of the feature will want "Only show selected" and "Hide selected" as options too. That could be implemented at the dataset-level via `fiftyone.utils.selection.select_objects()` and `fiftyone.utils.selection.hide_objects()`, but I think the more natural option is to implement it client-side and then perhaps add a (?) icon that links to documentation that explains how to use the `fiftyone.utils.selection` module.

I haven't fully thought it out, but managing the state of shown/hidden objects might get a bit complex. EG enabling a workflow like:
- select 5 objects -> only show selected
- then select 3 of the remaining objects -> hide selected
- then change samples and select 4 objects -> hide selected

Some initial thoughts on that point are:
- Since we allow selection across multiple samples, it may be a bit strange to do the filtering client-side 
- Since we

**Comment by brimoor** [comment]
> > Make the selection area for objects be the boundary of the box..., not the interior of the box
> 
> This is just a 1-2px difference, right? The current logic uses the actual coordinates of the object, while the line drawn around it may extend outside of the object a bit. This behavior might be possible to change, but I could add a tolerance of `line width / 2` to the boundary calculation fairly easily (I think), if that would do what you want.
> 
> > (including the label header)
> 
> Clicking on the title of a detection (e.g. "DOG" above) works for me - does it not work for you? Again, might be subject to a 1-2px difference depending on whether that header also has a border drawn.

Clicking on the header does work for me; yes. The only change is that clicking on the empty space inside a box should NOT select it, since most objects in crowded scenes are only selectable by clicking on their border that's within some other object's border.

Two notable things to consider here are (a) filled polylines, and (b) objects with instance masks. My "whatever is rendered is clickable" principle would indicate that the shaded interiors of those regions should be clickable. That already seems to be the behavior for filled polygons (eg testing on BDD100K dataset), but if that is too challenging to achieve for objects with instance masks, I think only the bbox being clickable would be intuitive enough.

**Comment by lethosor** [comment]
Oh, I understand what you're saying now. Any changes to that behavior would currently affect *all* boundary detection, i.e. any behavior that occurs when hovering over or clicking on objects. I'm a bit reluctant to implement exactly that behavior because it would make the click targets very small, and for non-overlapping boxes, it is convenient to have the box contents also be clickable. For instance, if the detection above actually contained a dog, users could reasonably expect to be able to click on the detected dog to select it, instead of having to aim for the box surrounding it. Regarding "whatever is rendered is clickable", the primary reasons we're not rendering anything in the middle of the box is so that users can see the contents of the box (or the instance mask, if there is one) - not that the box contains nothing that users would want to select.

I do agree that *preferring* to select boxes where the user is hovering over the border instead of just the interior would be a good user experience improvement, and I think I have an idea for how to implement that, although it could take some time (of which the new bounds calculations would probably be the most involved, and that would have been done in either case).

Making just instance masks clickable is an interesting idea - that might be possible, although the various drawing optimizations related to masks could prevent it from working or being accurate enough. I'll keep that in mind as a possible future enhancement.

**Comment by brimoor** [comment]
I think we have to go with boundary-only-style click/hover detection. It's not a matter of preference, it's just the reality that we can't design for non-overlapping boxes, because real data will have lots of overlap. Look at the images below for example.

Click targets might be small, but it's just not possible to do any meaningful object selection in the examples below if the box interiors are selectable. Some big box is going to cover up everything else and most things will be unselectable.

Screen Shot 2020-10-27 at 10 23 52 AM

Screen Shot 2020-10-27 at 10 21 35 AM

**Comment by lethosor** [comment]
What I am suggesting (and in fact, have mostly already implemented) is:
* When clicking on the border of a box, select that box
* Otherwise, if clicking inside a box, select that box

I think this covers the cases of both overlapping and non-overlapping boxes well enough for our purposes.

**Comment by brimoor** [comment]
Okay I didn't understand that from your previous comment; I like it.

**Comment by lethosor** [comment]
Got it. Player51 is updated on this branch and supports that now.

**Comment by benjaminpkane** [comment]
Can we clear selected objects when the dataset (and view?) changes?

https://github.com/voxel51/fiftyone/blob/37bd0467a63ebc94598684d7fba1fae69d920e14/fiftyone/core/session.py#L205

**Comment by lethosor** [comment]
Sure, thanks for the suggestion

File: electron/app/recoil/selectors.ts
Comment (benjaminpkane) [inline electron/app/recoil/selectors.ts:182]: What is the rationale here? This should probably be adjusted for numeric field range filters, as well.

I really want to clean up significant digit logic for these sliders. It just requires some time figuring out how to control the Material UI slider component. Or getting rid of it...

File: electron/app/utils/selection.ts
Comment (benjaminpkane) [inline electron/app/utils/selection.ts:3]: Loving the TypeScript

File: electron/app/components/SampleModal.tsx
Comment (benjaminpkane) [inline electron/app/components/SampleModal.tsx:342]: :eyes:

**Review by benjaminpkane — APPROVED** [review]
Very clean work here! Especially given the nuances.

Looks good to me from an implementation standpoint.

File: electron/app/recoil/selectors.ts
Comment (lethosor) [reply electron/app/recoil/selectors.ts:182]: I had an object with a random confidence (so ~10 decimal places), and the range was getting truncated to [0.52, 0.52] at some point, so the object was always getting filtered out (and I couldn't adjust the range because there was only one object).

There's likely a better place to do this, but I wasn't sure where - advice would be appreciated.

File: electron/app/components/SampleModal.tsx
Comment (lethosor) [reply electron/app/components/SampleModal.tsx:342]: For reference: I'm using a ref here so that updating the frame number doesn't re-render the entire modal. (There's some complicated logic around exactly how often timeupdate events fire, but there tend to be around 2-10 a second depending on load.)

**Review by lethosor — COMMENTED** [review]
The only things I think are left to do here are moving the dropdown and figuring out what to do about the filtering change I made (that you pointed out)

**Comment by lethosor** [comment]
Moved the dropdown:
image



> I'm pretty confident that users of the feature will want "Only show selected" and "Hide selected" as options too. That could be implemented at the dataset-level via `fiftyone.utils.selection.select_objects()` and `fiftyone.utils.selection.hide_objects()`, but I think the more natural option is to implement it client-side and then perhaps add a (?) icon that links to documentation that explains how to use the `fiftyone.utils.selection` module.
> 
> I haven't fully thought it out, but managing the state of hidden objects might get a bit complex. EG enabling a workflow like:
> 
> * select 5 objects -> only show selected
> * then select 3 of the remaining objects -> hide selected
> * then change samples and select 4 objects -> hide selected

@brimoor I think I missed this earlier - I agree that it would be useful to have these options, although I'm not sure about how feasible they are to implement on the client side, since the client-side filters don't support object IDs (in addition to state management being complicated like you described). I'll try for a server-side implementation for now and see how that works.

**Comment by brimoor** [comment]
> Moved the dropdown:
> image
> 
> > I'm pretty confident that users of the feature will want "Only show selected" and "Hide selected" as options too. That could be implemented at the dataset-level via `fiftyone.utils.selection.select_objects()` and `fiftyone.utils.selection.hide_objects()`, but I think the more natural option is to implement it client-side and then perhaps add a (?) icon that links to documentation that explains how to use the `fiftyone.utils.selection` module.
> > I haven't fully thought it out, but managing the state of hidden objects might get a bit complex. EG enabling a workflow like:
> > 
> > * select 5 objects -> only show selected
> > * then select 3 of the remaining objects -> hide selected
> > * then change samples and select 4 objects -> hide selected
> 
> @brimoor I think I missed this earlier - I agree that it would be useful to have these options, although I'm not sure about how feasible they are to implement on the client side, since the client-side filters don't support object IDs (in addition to state management being complicated like you described). I'll try for a server-side implementation for now and see how that works.

If things like the confidence/label filters are already implemented client-side, it seems like adding ID-filtering to the frontend is a natural extension to the abstraction there. But I'm not close to the details of the implementation and will certainly defer to you on how to best proceed right now.

**Comment by lethosor** [comment]
Turns out that it was easier to implement on the frontend after all. That's done in 318f5cd. That's likely all for this PR - tagging objects is a WIP on `obj-selection-tags`.

I'll merge tomorrow unless something else comes up here.

**Comment by brimoor** [comment]
Always being able to click on the title of boxes is sooooo helpful! I'm already finding this new UI super helpful. 💯

**Comment by brimoor** [comment]
One suggestion: for consistency with how label filters work, how about clearing selected objects when the modal closes? If I open the modal, apply a confidence filter, close and reopen, then the filter will be reset. It feels like the same should hold for selection as well.

**Comment by brimoor** [comment]
Now for a slightly larger suggestion regarding showing/hiding labels:

The ability to click -> hide labels is super useful. However, it's a bit confusing for information about hidden labels to be silently stored under the selected objects dropdown, with no visual indication that some labels may currently be hidden. I think a good solution for this is to store information about *hidden* labels within the label filter dropdown for the associated label field. EG if I click on some boxes and then select a hide option, the corresponding label filters will "activate" (light gray background) and when I expand the filters, I'll see a new section that looks similar to this:

```
Manual
[2 hidden x]

Labels
[+ add label]

Confidence
[slider]
``` 

I can then click the `x` on a per-label-field basis to unhide any manually hidden labels.

It's slightly more complex to implement this way, but I think it makes it very clear to the user what is happening.

Thoughts?

**Comment by brimoor** [comment]
One final thing: I think I like putting `[x objects selected]` underneath Display Options rather than inside the labels section. I think it may be a bit visually intimidating to see `[y selected]` and `[x objects selected]` so close together in the UI.

Also this hints at the fact that other types of things may be selectable in the future 😄

**Comment by lethosor** [comment]
Yeah, I agree the dropdown looks better here:
image
I left in support for adding things under individual headers in the sidebar, in case that's ever necessary.

I'm not familiar enough with the filter widgets to add a "manual" field, although that could be a better long-term solution (it's tricky because only the filtering logic has a special case for individual hidden objects currently - that information isn't stored as part of the filter object that the UI uses).

For now, I've made selected objects and hidden objects reset when the modal closes. I was originally reluctant to do this because closing the modal accidentally could result in users losing some important information, but it's likely fine either way for now (and this way prevents previous state from carrying over unexpectedly).

**Comment by brimoor** [comment]
Sweet, thanks for the updates! Approving now

**Review by brimoor — APPROVED** [review]
LGTM. Awesome new feature here

**Pull request merged**
